File: 2nd_quarter/week01/Whaeun/주차 요금 계산.cpp
#include <string>
#include <vector>
#include <map>
#include <sstream>
#include <cmath>
#include <algorithm>

using namespace std;

// calculate parking time
int parkingTime(string in, string out) {
    
    int hour1 = stoi(in.substr(0, 2));
    int minute1 = stoi(in.substr(3, 2));
    int hour2 = stoi(out.substr(0, 2));
    int minute2 = stoi(out.substr(3, 2));
    
    // calculate as minute
    int parking_time = (hour2-hour1)*60 + (minute2-minute1);
    
    return parking_time;
}


vector<int> solution(vector<int> fees, vector<string> records) {
    vector<int> answer;
    vector<pair<int, int>> num_answer;
    
    map<string, vector<string>> record;
    
    stringstream ss;
    
    for(int i = 0; i < records.size(); i++){
        string time, num, status; 
        
        ss.str(records[i]);
        
        ss >> time >> num >> status;
        
        //mapping time with car_num
        record[num].push_back(time);
        
        ss.clear();
        
    }
    
    for(auto iter: record) {
        
        // doesn't have leaving time
        if(iter.second.size() % 2 == 1){
            iter.second.push_back("23:59");
        }
        
        
        int sum = 0;
        
        // calculate each parking fee
        for(int i = 0; i < iter.second.size() - 1; i += 2) {
            sum += parkingTime(iter.second[i], iter.second[i+1]);
        }
        
        // basic fee
        int price = fees[1];
        
        // over basic fee 
        if(sum > fees[0]) {
            price += ceil((double)(sum-fees[0]) / (double)fees[2]) * fees[3];
        }
        
        answer.push_back(price);
    }
    
    return answer;
}

File: 2nd_quarter/week01/Royce/주차 요금 계산.py
import math
def cal(m, fees):
    return fees[1] + math.ceil(max(0, (m - fees[0])) / fees[2]) * fees[3]

def solution(fees, records):
    answer = []
    
    park = dict()
    arr = dict()
    
    for k in records:
        time, car, status = k.split()
        h, m = time.split(":")
        m = int(h) * 60 + int(m)
        
        if status == "IN":
            park[car] = m
        elif status == "OUT":
            try:
                arr[car] += m - park[car]
            except:
                arr[car] = m - park[car]
            del park[car]
    
    # items는 (key, value)를 뜻하고, key, value는 따로 해준다.
    # 남은 것들에 대하여 최대 요금으로 때리기
    for car, m in park.items():
        try:
            arr[car] += 23*60+59 - m
        except:
            arr[car] = 23*60+59 - m
            
    return [cal(time, fees) for car, time in sorted(list(arr.items()), key = lambda x : x[0])]
# time에 대해서 정렬해서 fee를 넣어줌
    

# 시도했다가 아예 틀림..
# def solution(fees, records):
#     answer = []
    
#     # 기본 시간, 기본 요금, 단위 시간, 단위 요금
#     def cal(n):
#         if n <= fees[0]:
#             return fees[1]
#         else:
#             return (((n - fees[0]) // fees[2]) + 1) * fees[3]
    
#     arr = {}
#     for k in records:
#         come = int(k[:2]) * 60 + int(k[3:5])
#         arr[k[6:10]] = come
    
#     return answer

